### ISSUE TITLE
Warn if an == implementation does not check runtimeType in generic classes

### ISSUE THREAD

**Issue by rrousselGit** [dart-lang/linter#4706]
Hello!

A common pattern with == is to write;

```dart
bool operator ==(Object other) {
  return other is MyClass && other.value == value;
}
```

That works fine, but it starts failing with generic classes. If we write:

```dart
bool operator ==(Object other) {
  return other is MyClass<T> && other.value == value;
}
```

Then this can break the == protocol as `a == b` might not be equal to `b == a`. Because `MyClass<int> is MyClass<num>` works, but `MyClass<num> is MyClass<int>` does not.
The solution is generally to check runtimeType.

A warning could check its presence in the == implementation of generic classes, and warn if missing.

**Comment by srawlins** [comment]
> That works fine, but it starts failing with generic classes. If we write:

Don't non-generic classes also contain the same non-commutative problem? I guess only if `operator ==` is overridden in a subclass as well.

**Comment by rrousselGit** [comment]
True there's subclassing too. That'd probably involve lots of false positives though, for classes which are not subclassed.

**Comment by lrhn** [comment]
Are we sure that type equality is the correct choice, or would mutual subtyping be a better choice?

Otherwise `MyClass<dynamic>` and `MyClass<Object?>` will be unequal, even though they're practically indistinguishable at runtime.

That does take more code to check, double dispatch probably:
```dart
bool operator==(Object other) =>
    other is MyClass<T> && other._eq(this);
bool _eq(MyClass first) =>
     first is MyClass<T> && first.value == value;
```
And you probably have to omit the type argument from the hash code.

But it's a better extensional equally.

**Comment by rrousselGit** [comment]
@lrhn Wouldn't that still allow subclassing?
runtimeType checking has the benefit of rejecting subclasses

**Comment by lrhn** [comment]
It would allow subclasses that do not override `operator==` (and `_eq`) to still be equal, and allow those who do override them to be unequal. 

In that way, it's better than checking equality of the runtime type, which doesn't allow any other class to be equal, even if it wants to, and deserved to be considered equal. 
If you want to allow that to other libraries too, of probably make `_eq` public and give it the full responsibility for equality, like

```dart
bool operator==(Object other) 
  => other is MyClass && other.equalsFrom<T>(this);
other.sameMyClassValues<T>(this);
bool equalsFrom<S>(MyClass<S> first) =>
    this is MyClass<S> && first is MyClass<T> && sameValues<S, T>(first, this);

static bool sameValues<S, T>(MyClass<S> first, MyClass<T> other) =>
    this.value == other.value;
```
That way any subclass can decide which types it wants to be equal to, and can define its own `sameValues`, which can call the super-type's `sameValues` for the original fields.

And you can create mocks.

**Comment by rrousselGit** [comment]
> It would allow subclasses that do not override operator== (and _eq) to still be equal, and allow those who do override them to be unequal.

That sounds like a niche scenario. In my opinion, most cases don't want this.

I believe most common case is likely to check runtimeType and annotate the == with `@mustBeOverridden`, such that subclasses should override == too (where they could invoke super.== if they wish)
